source/blog/_posts/2018-01-12-mutation-testing-with-infection.md

```diff
@@ -0,0 +1,374 @@
+Tests: 76 Line Coverage: 69.15%
+
+Humbug is analysing source files...
+
+Mutation Testing is commencing on 32 files...
+(.: killed, M: escaped, S: uncovered, E: fatal error, T: timed out)
+
+......M.....M..SSS...................M.......M..M......M.M.. |   60 (17/32)
+MMM.MS.....SSM.M...T........TSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS |  120 (25/32)
+SSSSSSSS.....M......MSSS...M....M.M.
+
+156 mutations were generated:
+      88 mutants were killed
+      48 mutants were not covered by tests
+      18 covered mutants were not detected
+       0 fatal errors were encountered
+       2 time outs were encountered
+
+Metrics:
+    Mutation Score Indicator (MSI): 58%
+    Mutation Code Coverage: 69%
+    Covered Code MSI: 83%
+
+Time: 53.84 seconds Memory: 8.00MB
+Humbug results are being logged as TEXT to: humbuglog.txt
+```
+
+Next I had to create a small script to have the possibility to take the execution time of infection. No rocket science but it does the job.
+
+```bash 
+START=`date +%s%N`
+
+./bin/infection
+
+END=$((`date +%s%N` - $START))
+
+bc <<< "scale=2; $END/1000000000"
+```
+
+Let us start infection and take the time. In the first run infection needed overall ~ 42 seconds and created 223 mutations. As we can see infection creates more mutations and needs less time. Awesome :-) 
+
+```bash
+   ____      ____          __  _
+   /  _/___  / __/__  _____/ /_(_)___  ____ 
+   / // __ \/ /_/ _ \/ ___/ __/ / __ \/ __ \
+ _/ // / / / __/  __/ /__/ /_/ / /_/ / / / /
+/___/_/ /_/_/  \___/\___/\__/_/\____/_/ /_/
+ 
+Running initial test suite...
+
+Phpunit version: 6.5.5
+
+   82 [============================] 2 secs
+
+Generate mutants...
+
+Processing source code files: 32/32
+Creating mutated files and processes: 223/223
+.: killed, M: escaped, S: uncovered, E: fatal error, T: timed out
+
+....TMM.....ME.ES..S............................E.   ( 50 / 223)
+................M....M.M..M....EE...E....MMM.SM..M   (100 / 223)
+MMM.M.SS............SSSM...T......TSSSSSSSSSSSSSSS   (150 / 223)
+SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS....S...M..   (200 / 223)
+.......M......M.M......                              (223 / 223)
+
+223 mutations were generated:
+     130 mutants were killed
+      63 mutants were not covered by tests
+      21 covered mutants were not detected
+       6 errors were encountered
+       3 time outs were encountered
+
+Metrics:
+         Mutation Score Indicator (MSI): 62%
+         Mutation Code Coverage: 72%
+         Covered Code MSI: 87%
+
+Please note that some mutants will inevitably be harmless (i.e. false positives).
+42.33    
+```
+
+At last we should test the threading option of inception. Be aware this option can give you many false-positives results if your tests depends on each other or use a non stateless database for testing purpose. I started infection with 4 threads and the result is amazing. Instead of 42 Seconds infection needs only 19 seconds to execute the whole mutation stuff.
```

Comment:
do you mean infection instead of inception? :smile:

Reply:
yes - :joy: